---
title: "Correlated direct-indirect genetic effects with transmitted/non-transmitted PRS simulations"
author: "Mark Adams"
output: html_notebook
---

```{r}
library(mvtnorm)
```

# Model

The phenotype $P$ is defined as

\[
\mathrm{P} = \mathrm{A} + \mathrm{IGE}_\mathrm{maternal} + \mathrm{IGE}_\mathrm{paternal} + \mathrm{E} 
\]

where

- $\mathrm{A}$ is the additive genetic effect
- $\mathrm{IGE}$ is the indirect genetic effect from the mother and father. $\mathrm{IGE}_\mathrm{maternal} + \mathrm{IGE}_\mathrm{paternal}$ can be considered the genetic nurture effect.
- $\mathrm{E} is the individual environment effect

The variance component model we are simulating is therefore:

\[
V_\mathrm{P} = V_\mathrm{A} + 2V_\mathrm{IGE} + V_\mathrm{E}
\]

where

- $V_\mathrm{P}$ is the phenotypic variance
- $V_\mathrm{A}$ is the direct genetic effect variance (additive genetic variance)
- $V_\mathrm{IGE}$ is the indirect genetic effect variance (effect of parents' genes via the environment on their offspring). The contribution of the variance is doubled because there is an effect from each parent
- $V_\mathrm{E}$ is the environment variance

and where $V_\mathrm{A}$ and $V_\mathrm{IGE}$ are correlated 

\[
\Sigma = \begin{bmatrix} V_\mathrm{A}       & COV_\mathrm{A,IGE} \\
                         COV_\mathrm{A,IGE} & V_\mathrm{IGE}
          \end{bmatrix}
\]

We also simulate a polygenic score (PRS) that is an estimate of $\mathrm{A}$ based on GWAS data and define it based on the proportion of $V\mathrm{A}$ that it captures. As a proxy of $\mathrm{IGE}$, we split the PRS into "transmitted" ($\mathrm{T}$) and "nontransmitted" ($\mathrm{NT}$) scores, so we define 

\[
\mathrm{A} = \mathrm{T} + \mathrm{NT}
\]
       
# Parameters

Define the basic parameters for the simulation. 

```{r}
V_P = 1 # phenotypic variance
h2 = 0.5 # heritability (direct genetic effect proportion) (V_A/V_P)
ige2 = 0.1 # indirect genetic effect proportion (2*V_IGE/V_P)
s2 = 0.05 # proportion of V_A captured by polygenic scores
r_a_ige = 0.3 # genetic correlation between direct and indirect effect
n <- 100000 # sample size (number of trios)
```

The remainder of the parameters can be defined from these.

Variance components:

```{r}
V_A <- h2*V_P # additive genetic variance
V_T <- V_A/2 # transmitted genetic variance
V_NT <- V_A/2 # non-transmitted genetic variance
V_IGE <- ige2*V_P/2 # indirect genetic variance
V_PRS <- s2*V_A # PRS variance
```

which gives us

```{r}
V_A
V_T
V_NT
2*V_IGE
V_PRS
```

We also need the covariance between $V_\mathrm{A}$ and $V_\mathrm{IGE}$, $COV_\mathrm{A,IGE}$. Since the correlation is 

\[
r_\mathrm{a,ige} = \frac{COV_\mathrm{A,IGE}}{\sqrt{V_\mathrm{A} \times V_\mathrm{IGE}}}
\]

then

\[
COV_\mathrm{A,IGE} = r_\mathrm{a,ige} \times \sqrt{V_\mathrm{A}\times V_\mathrm{IGE}}
\]

```{r}
COV_A_IGE = r_a_ige * sqrt(V_A * V_IGE)
COV_A_IGE
```

The transmitted and non-transmitted portions of the direct genetic effect will have half the covariance with the indirect genetic effect
```{r}
COV_T_IGE <- COV_A_IGE/2
COV_NT_IGE <- COV_A_IGE/2
```

PRS has a correlation of `r s2` with the additive genetic variance and has a transmitted and non-transmitted component as well.

```{r}
V_T_PRS <- V_PRS/2
V_NT_PRS <- V_PRS/2
```

The covariance of the transmitted and non-transmitted PRS with the total genetic effects:

\[
s = \frac{COV_\mathrm{T,PRS_T}}{\sqrt{V_TV_\mathrm{PRS_T}}}
\]
Rearrange:

\[
COV_\mathrm{T,PRS_T} = s \sqrt{V_TV_\mathrm{PRS_T}}
\]

Substitute value of $s^2 = V_\mathrm{PRS_T} / V_T$

\[
\begin{aligned}
COV_\mathrm{T,PRS_T} & = \sqrt{\frac{V_\mathrm{PRS_T}}{V_T}}\sqrt{V_TV_\mathrm{PRS_T}} \\
& = \sqrt{V_\mathrm{PRS_T}} \sqrt{V_\mathrm{PRS_T}} \sqrt{\frac{1}{V_T}}\sqrt{V_T} \\
& = V_\mathrm{PRS_T}
\end{aligned}
\]


The covariance of the transmitted and non-transmitted PRS with the indirect genetic effect:

```{r}
COV_T_IGE <- COV_A_T * COV_A_IGE
COV_NT_IGE <- COV_A_NT * COV_A_IGE
COV_T_PRS_IGE <- COV_A_PRS * COV_T_IGE
COV_NT_PRS_IGE <- COV_A_PRS * COV_NT_IGE
COV_T_IGE
COV_NT_IGE
COV_T_PRS_IGE
COV_NT_PRS_IGE
```

The genetic effects variance-covariance matrix is a 4-by-4 matrix. Transmitted and non-transmitted effects will be uncorrelated with each other.                 
                  
```{r}
genetic_effect_names <- 
                 c('A_T',         'PRS_T',    'A_NT',        'PRS_NT',       'indirect')
Sigma = matrix(c(    V_T,         V_T_PRS,         0,               0,       COV_T_IGE,
                 V_T_PRS,         V_T_PRS,         0,               0,   COV_T_PRS_IGE,
                       0,               0,       V_NT,       V_NT_PRS,      COV_NT_IGE,
                       0,               0,   V_NT_PRS,       V_NT_PRS,  COV_NT_PRS_IGE,
                 COV_T_IGE, COV_T_PRS_IGE, COV_NT_IGE, COV_NT_PRS_IGE,           V_IGE),
             nrow=5, ncol=5, byrow=T, dimnames=list(genetic_effect_names, genetic_effect_names))
Sigma
```

When determining the residual variance for the environment effect, the increase in variance from the covariance between the direct and indirect effects also needs to be subtracted. The variance of the sum of two variables is $\mathrm{var}(X+Y) = \mathrm{var}(X) + \mathrm{var}(Y) - 2\mathrm{cov}(X, Y)$. While there are two $V_\mathrm{IGE}$ components contributing (one from each parent) we only need to remove $COV_\mathrm{A,IGE}$ because the covariance we defined above is that between a target's direct genetic effect on themselves and the indirect effect on the target's offspring but the contribution of this covariance to the phenotypic variance is between a target's direct genetic effect on the target's phenotype and the parents' (summed) indirect effect on the target's phenotype

```{r}
V_E = V_P - V_A - 2*V_IGE - COV_A_IGE
V_E
```

# Simulation

For simulating data from the covariance matrix, the function `rmvnorm` is the multivariate equivalent of `rnorm`, parameterised by a vector of means and a covariance matrix `sigma`. Simulate genetic effects for mothers and fathers

```{r}
maternal_genetic <- data.frame(rmvnorm(n, sigma=Sigma))
paternal_genetic <- data.frame(rmvnorm(n, sigma=Sigma))
names(maternal_genetic) <- names(paternal_genetic) <- genetic_effect_names
```

This yields matrices of size $n \times 5$, so the first and second columns are the transmitted genetic and PRS effects, the third and fourth columns are the nontransmitted effects, and the fifth column is the indirect effect

```{r}
head(maternal_genetic)
```

Note that the indirect effects here are the genetic nurture effects on the phenotype of the offspring.

Simulate the other components of the parental phenotypes. Since we are simulating the grandparental indirect effect separately from the parental direct effects, we need to make up the missing variance by adding their covariance term into the residual (environmental )

```{r}
maternal_genetic_nurture <- rnorm(n, sd=sqrt(2*V_IGE)) # mother's parents' (combined) indirect effects (on the mother's phenotype)
paternal_genetic_nurture <- rnorm(n, sd=sqrt(2*V_IGE)) # father's parents' (combined) indirect effects (on the mother's phenotype)
maternal_environment <- rnorm(n, sd=sqrt(V_E+COV_A_IGE))
paternal_environment <- rnorm(n, sd=sqrt(V_E+COV_A_IGE))
```

Add up components to create parental phenotypes

```{r}
maternal_genetic_direct <- maternal_genetic$A_T + maternal_genetic$A_NT
maternal_phenotype <- maternal_genetic_direct + maternal_genetic_nurture + maternal_environment
paternal_genetic_direct <- paternal_genetic$A_T + paternal_genetic$A_NT
paternal_phenotype <- paternal_genetic_direct + paternal_genetic_nurture + paternal_environment
```

Check that the variances of the parental phenotpyes are as expected $\approx V_\mathrm{P}$ = `r V_P`.
```{r}
var(maternal_phenotype)
var(paternal_phenotype)
```

The transmitted and non-transmitted PRS add up to the total parental PRS
```{r}
maternal_prs <- maternal_genetic$PRS_T + maternal_genetic$PRS_NT
paternal_prs <- paternal_genetic$PRS_T + paternal_genetic$PRS_NT
```

Create offspring components

```{r}
offspring_genetic_direct <- maternal_genetic$A_T + paternal_genetic$A_T  # transmitted effects
offspring_genetic_nurture <- maternal_genetic$indirect + paternal_genetic$indirect # indirect genetic effects of parents on offspring
offspring_environment <- rnorm(n=n, sd=sqrt(V_E))
offspring_phenotype <- offspring_genetic_direct + offspring_genetic_nurture + offspring_environment
offspring_prs <- (maternal_genetic$PRS_T + paternal_genetic$PRS_T) # PRS from parental transmitted PRS values
```


```{r}
var(offspring_genetic_direct)
var(offspring_genetic_nurture)
var(offspring_environment)
var(offspring_phenotype)
var(offspring_prs)
```

